h/views/oidc.py

```diff
@@ -183,49 +260,61 @@ def redirect(self):
             # You must be logged out in order to log in.
             raise HTTPForbidden
 
-        # Get the user's ORCID iD from ORCID.
-        orcid_id = self._orcid_client.get_orcid(validated_params["code"])
+        # Get the user's provider unique ID from the provider.
+        provider_unique_id = self._oidc_service.get_provider_unique_id(
+            self.settings.provider, validated_params["code"]
+        )
 
         # Get the existing Hypothesis account that's already connected to this
-        # ORCID iD, if any.
-        user = self._user_service.fetch_by_identity(IdentityProvider.ORCID, orcid_id)
+        # provider unique ID, if any.
+        user = self._user_service.fetch_by_identity(
+            self.settings.provider, provider_unique_id
+        )
+
+        action = decoded_state.action
 
-        actions = {
-            "connect": self.connect_orcid_id,
-            "login": self.log_in_with_orcid,
-        }
-        action_method = actions[decoded_state.action]
-        return action_method(orcid_id, user)
+        match action:
+            case "connect":
+                return self.connect_provider_unique_id(provider_unique_id, user)
+            case "login":
+                return self.log_in_with_provider(provider_unique_id, user)
+            case _:
+                raise UnexpectedActionError(action)
```

Comment:
I've tweaked this to use `match` instead of dictionary-dispatch and to raise `UnexpectedActionError` instead of `KeyError` if the action is not `"connect"` or `"login"`. Of course this should never happen, but it could happen for example if future code changes state generating `state` JWTs with a new action value but forget to update the `match` here.